File: Stack & Queue/Find the Most Competitive Subsequence/solution_segmentTree.ts
// O(n + klogn)
function mostCompetitive(nums: number[], k: number): number[] {
    let result = []
    let segmentTree = new SegmentTree(0, nums.length - 1, 0)
    for(let i = 0; i < nums.length; i++) {
        segmentTree.update(i, i, nums[i])
    }
    function helper(startIndex: number, k: number) {
        if(k === 0) return
        let min = segmentTree.query(startIndex, nums.length - k)
        result.push(min[0])
        helper(min[1] + 1, k - 1)
    }
    helper(0, k)
    return result
};

class SegmentTree {
    start
    end
    left
    right
    min
    index
    
    constructor(start, end, min) {
        this.start = start
        this.end = end
        this.min = min
    }
    
    query(start, end) {
        if(this.start > end || this.end < start) {
            return [Number.MAX_SAFE_INTEGER, - 1]
        }
        
        if(this.start >= start && this.end <= end) {
            return [this.min, this.index]
        }
        this.down()
        const left = this.left.query(start, end)
        const right = this.right.query(start, end)
        if(left[0] <= right[0]) {
            return left
        } else {
            return right
        }
    }
    
    update(pos, index, value) {
        if(this.start > pos || this.end < pos) {
            return
        }
        if(this.start === this.end) {
            this.min = value
            this.index = index
            return
        }
        this.down()
        this.left.update(pos, index,value)
        this.right.update(pos, index, value)
        if(this.left.min <= this.right.min) {
            this.min = this.left.min
            this.index = this.left.index
        } else {
            this.min = this.right.min
            this.index = this.right.index
        }
        
    }
    
    down() {
        if(this.start !== this.end) {
            if(!this.left || !this.right) {
                const mid = this.start + Math.floor((this.end - this.start)/ 2)
                this.left = new SegmentTree(this.start, mid, this.min)
                this.right = new SegmentTree(mid + 1, this.end, this.min)
            }
        }
    }
}

File: Design/Range Sum Query - Mutable/solution_segment_tree.py
class SegmentTree:
    def __init__(self, start, end):
        self.start = start
        self.end = end
        self.val = 0
        self.left, self.right = None, None

    def query(self, start, end):
        if start > self.end or end < self.start:
            return 0

        if start <= self.start and end >= self.end:
            return self.val

        self.down()
        return self.left.query(start, end) + self.right.query(start, end)

    def update(self, start, end, val):
        if start > self.end or end < self.start:
            return

        if start <= self.start and end >= self.end:
            self.val = val
            return

        self.down()
        self.left.update(start, end, val)
        self.right.update(start, end, val)
        self.val = self.left.val + self.right.val

    def down(self):
        if self.start < self.end:
            if not self.left:
                mid = self.start + (self.end - self.start) // 2
                self.left = SegmentTree(self.start, mid)
                self.right = SegmentTree(mid + 1, self.end)

class NumArray:

    def __init__(self, nums: List[int]):
        self.st = SegmentTree(0, len(nums) - 1)
        for i, num in enumerate(nums):
            self.st.update(i, i, num)

    def update(self, index: int, val: int) -> None:
        self.st.update(index, index, val)

    def sumRange(self, left: int, right: int) -> int:
        return self.st.query(left, right)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)
